20241016-R-error-code-metrics.md

```diff
@@ -0,0 +1,111 @@
+
+## Background
+
+Currently, error codes are only visible as logs when Dapr runs in `debug` log mode, which is impractical for both production apps and observability. 
+
+Additionally, the error message is a field in JSON response body as such:
+```
+{
+	...
+    "errorCode": "ERR_STATE_GET",
+    ...
+}
+```
+...which means observation of these codes requires increased log noise via the `debug` flag, as well as parsing the logs to find this field. There also wouldn't be automatic data retention, so logs would have to be manually stored to provide a historical view.
+
+
+Dapr uses Prometheus by default, and so having these errors as Prometheus metrics will allow querying, aggregation, and data retention of the codes.
+
+## Related Items
+
+Initial Error Codes Proposal: [proposals/#21](https://github.com/dapr/proposals/pull/21)
+
+Error Codes Implementation Parent: [dapr/#6068](https://github.com/dapr/dapr/issues/6068)
+
+## Expectations and alternatives
+
+### Expectations
+- When opted-in, it would add up to  `N * app`  (N = number of error codes) metrics for scraping, today that's around `52 * app`.
+- If a component-based error, codes could repeat for multiple components used by the same app, and wouldn't be differentiated by the metrics, which only provide the `app-id`.
+- Many logged error codes are currently defined in [pkg/messages/predefined.go](https://github.com/dapr/dapr/blob/master/pkg/messages/predefined.go), but not all. Implementation requires case-by-case analysis of usage of error codes in that uppercase format `ERR_NAME_TAG`.
+
+### Alternatives
+- One alternative solution which retains error codes as logs only would be to change the level of logs containing these error codes from `debug` to `info`. However, this increases noise and raises logs which may be unhelpful at `info` level.
+- Another alternative solution would be to log error codes separately with opt-in functionality, but this achieves the same result as adding metrics, but increase log noise and doesn't introduce a simple way to actually work with the errors.
+
+## Implementation Details
+
+### Solution
+
+The feature is enabled via an opt-in flag within Dapr HTTP metrics (disabled by default). The goal of the opt-in functionality is to prevent the increased cost of additional metrics if they aren't being used.
+
+```yaml
+spec:
+  metric:
+    enabled: true
+	recordErrorCodes: true # <-- new opt-in flag
+	http:
+		...
+	...
+...
+```
+From the user's perspective, the metric `dapr_error_code_count` could be split by `app-id` and `error_code` like so:
+```go
+dapr_error_code_count{app_id="service-a", error_code="ERR_ACTOR_RUNTIME_NOT_FOUND"} 4
+dapr_error_code_count{app_id="service-a", error_code="ERR_ACTOR_INVOKE_METHOD"} 12
+dapr_error_code_count{app_id="service-b", error_code="ERR_ACTOR_RUNTIME_NOT_FOUND"} 55
+dapr_error_code_count{app_id="service-b", error_code="ERR_DIRECT_INVOKE"} 2
```

Comment:
In addition to tracking error codes, I believe users would benefit from having a type label (or similar) to categorize the errors by API or functional area. This would allow users to filter error metrics more efficiently without needing to hardcode specific error codes on their side. For example, users could retrieve all actor-related errors or pubsub-related errors for a particular app with ease.

An example of this would be:
```
dapr_error_code_count{app_id="service-b", error_code="ERR_ACTOR_RUNTIME_NOT_FOUND", type="ACTORS_API"} 55
```

Possible categories for the type label could include actors, pubsub, state, secrets, health, scheduler, operator, etc. This could improve the granularity of error tracking and make it easier to manage metrics across different Dapr APIs.

Reply:
I agree that we would be providing a better developer experience by making it possible to group errors by API.
We do follow a schema in naming the error codes, so for example a prefix search on `ERR_ACTOR_RUNTIME*` would only return errors from the actor runtime, but this is not enforced and until now has been left to the developer as a convention.
We'd have to change the code to properly track the API. I already suggested something that would go in that direction in [this comment](https://github.com/dapr/dapr/pull/8192#pullrequestreview-2382399008).

Reply:
I agree with @nelson-parente's point here, I think it would be useful to add a `type` or `group` for each error code, as this would allow easier querying and filtering for all Dapr metrics esp in backends. @jake-engelberg  can you add this to the proposal?

Reply:
I have updated the proposal to account for aligning the error names to allow filtering by prefix, as well as general improvements to the implementation explanation